lib/compat/wordpress-6.7/class-wp-block-templates-registry.php

```diff
@@ -0,0 +1,269 @@
+<?php
+/**
+ * Templates registry functions.
+ *
+ * @package Gutenberg
+ * @since 6.7.0
+ */
+
+if ( ! class_exists( 'WP_Block_Templates_Registry' ) ) {
+	/**
+	 * Core class used for interacting with templates.
+	 *
+	 * @since 6.7.0
+	 */
+	final class WP_Block_Templates_Registry {
+		/**
+		 * Registered templates, as `$name => $instance` pairs.
+		 *
+		 * @since 6.7.0
+		 * @var WP_Block_Template[] $registered_block_templates Registered templates.
+		 */
+		private $registered_templates = array();
+
+		/**
+		 * Container for the main instance of the class.
+		 *
+		 * @since 6.7.0
+		 * @var WP_Block_Templates_Registry|null
+		 */
+		private static $instance = null;
+
+		/**
+		 * Registers a template.
+		 *
+		 * @since 6.7.0
+		 *
+		 * @param string $template_name Template name including namespace.
+		 * @param array  $args          Optional. Array of template arguments.
+		 * @return WP_Block_Template|false The registered template on success, or false on failure.
+		 */
+		public function register( $template_name, $args = array() ) {
+
+			$template = null;
+
+			if ( ! is_string( $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Template names must be a string.', 'gutenberg' ),
+					'6.7.0'
+				);
+				return new WP_Error( 'template_name_no_string', __( 'Template names must be a string.', 'gutenberg' ) );
+			}
+
+			if ( preg_match( '/[A-Z]+/', $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Template names must not contain uppercase characters.', 'gutenberg' ),
+					'6.7.0'
+				);
+				return new WP_Error( 'template_name_no_uppercase', __( 'Template names must not contain uppercase characters.', 'gutenberg' ) );
+			}
+
+			$name_matcher = '/^[a-z0-9-]+\/\/[a-z0-9-]+$/';
+			if ( ! preg_match( $name_matcher, $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Template names must contain a namespace prefix. Example: my-plugin//my-custom-template', 'gutenberg' ),
+					'6.7.0'
+				);
+				return new WP_Error( 'template_no_prefix', __( 'Template names must contain a namespace prefix. Example: my-plugin//my-custom-template', 'gutenberg' ) );
+			}
+
+			if ( $this->is_registered( $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					/* translators: %s: Template name. */
+					sprintf( __( 'Template "%s" is already registered.', 'gutenberg' ), $template_name ),
+					'6.7.0'
+				);
+				/* translators: %s: Template name. */
+				return new WP_Error( 'template_already_registered', sprintf( __( 'Template "%s" is already registered.', 'gutenberg' ), $template_name ) );
```

Comment:
I'd consider making this more DRY, and piping the `WP_Error` message into the `_doing_it_wrong` notice.

Reply:
I refactored it in 90acfe8d2ff48398bbb7222e79b15898e94655a9. Let me know how it looks.